RULES/cc/cc.go

```diff
@@ -204,11 +178,33 @@ func (lib Library) Build(ctx core.Context) {
 
 // CcLibrary for Library returns the library itself, or a toolchain-specific variant
 func (lib Library) CcLibrary(toolchain Toolchain) Library {
-	toolchain = toolchainOrDefault(toolchain)
+	if toolchain == nil {
+		core.Fatal("CcLibrary() called with nil toolchain.")
+	}
+
+	if lib.Out == nil {
+		core.Fatal("Out field is required for cc.Library")
+	}
 
-	if !ToolchainAccepts(toolchain, toolchainOrDefault(lib.Toolchain)) {
-		core.Fatal("Library %s does not support toolchain %s", lib.Out.Relative(), toolchain.Name())
```

Comment:
ToolchainAccepts is currently used for testing toolchains in an internal repo. I have sent you a private message with the link to the use.

Reply:
Oh, I see. Previously if you specified a toolchain in a rule, it was an error if you tried to build it with another toolchain (unless that parent toolchain accepts the toolchain in the rule). The change here just makes it always ok.

I think that's fine... it just means dbt will now not catch some build configuration errors, for example, trying to link an ARM-specific kernel library (with a specified toolchain) into an X64 binary. That doesn't seem horrible, since I guess the linker fails anyway.

Note that the libsupcxx/boot build rules (the only(?) source of explicit toolchains in library targets) have explicit toolchains because they need to change the standard dependencies, and the only way to do that is to specify a complete toolchain. It may be nicer if the Toolchain field instead of specifying a toolchain, instead specified a function that takes and returns a (Toolchain, error). The libsupcxx rules could then just modify the stddeps of the passed-in toolchain, and testing binary targets could modify a toolchain to remove optimization parameters, add sanitizer options etc.

I mention that design choice here, because then there's no question of what it means to build a target with a "wrong" toolchain.

Reply:
> Previously if you specified a toolchain in a rule, it was an error if you tried to build it with another
> toolchain (unless that parent toolchain accepts the toolchain in the rule). The change here just makes it always ok.

Yes exactly. I think that it's worth to make the dbt-rules logic simpler and offload the checking to linker.
 
> It may be nicer if the Toolchain field instead of specifying a toolchain, instead specified a function that
> takes and returns a (Toolchain, error).

What you're proposing is similar to CMake's PUBLIC `target_compile_options()`. It allows adding flags for a specific target. With PRIVATE modifier, only the target is affected, with PUBLIC also the dependencies. The current `CompilerFlags` field is thus equivalent of the PRIVATE modifier and changing the toolchain, since it propagates to dependencies, would be equivalent of the PUBLIC one. Note that here I'm assuming that you meant for the toolchain changes to be propagated.

Right now all CC target dependencies are compiled with the same toolchain so this would be a major change. We'd need to append (a hash of?) flags and other toolchain state to its name to distinguish between the same toolchain with different options.

So basically this boils down to -- will we ever need to use more than one toolchain for a single target and all its dependecies? If the answer is yes, then we should figure out a way to support this.

What I don't like right now is that when I want to add a recursive "-g" option to the target for debugging, I have to change the toolchain itself in its sources. We should figure out a way to support things like this, either via above Paul's proposal, or build modes (Release, Debug, ...), or something different.

**exp-baremetal-math use case**
*(For context see [exp-baremetal-math](https://p.daedalean.ai/diffusion/EXPBAREMETALMATH/browse/master/) testing setup)*
For your usecase Paul, I think the above is not really needed and to avoid specifying `cc-toolchain=linux-x86_64-libsupcxx-test` in the command line every time, you can make a function in `RULES/cctest/cctest.go` like `makeTest()` which takes sources, dependencies, adds the toolchain, and returns `cc.Binary`. Then each test in `BUILD.go` would be specified as `var MyTest = makeTest(...)`.

Reply:
I agree with most of what you say - it seems like a good summary. But one thing that I think you've missed it that currently NOT all CC target dependencies are compiled with the same toolchain -- in libsupcxx there's a toolchain that's used specifically for compiling the bottom-level dependencies (called bootToolchain). It has the same name as the top-level toolchain so that dbt doesn't notice that it's different. That was one of the motivations for the suggestion above about toolchain modification.